### PR TITLE
AP_Camera: Load CAMERA_INFORMATION and VIDEO_STREAM_INFORMATION from Lua

### PR DESCRIPTION
**Pull request by nexton-winjeel**
This PR allows us to populate the [`CAMERA_INFORMATION`](https://mavlink.io/en/messages/common.html#CAMERA_INFORMATION) and [`VIDEO_STREAM_INFORMATION`](https://mavlink.io/en/messages/common.html#VIDEO_STREAM_INFORMATION) MAVLink messages from Lua scripts. 

Sending these messages allows a GCS to auto-configure the video stream to receive video from the payload.

Replaces https://github.com/ArduPilot/ardupilot/pull/27594.

### CODE REVIEW

**Comment by amilcarlucas** [comment]
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   3516 (+0.2642%)  0 (0.0000%)  4 (+0.0015%)   3516 (+0.2638%)                                  629700
arduplane        3436 (+0.1915%)  0 (0.0000%)  4 (+0.0015%)   3436 (+0.1911%)                                  164640
blimp            3520 (+0.2602%)  0 (0.0000%)  0 (0.0000%)    3520 (+0.2598%)                                  607480
arducopter-heli  3428 (+0.1908%)  0 (0.0000%)  -4 (-0.0015%)  3428 (+0.1904%)                                  162216
arducopter       3428 (+0.1907%)  0 (0.0000%)  4 (+0.0015%)   3428 (+0.1903%)                                  161344
ardusub          3432 (+0.2158%)  0 (0.0000%)  0 (0.0000%)    3432 (+0.2155%)                                  370092
ardurover        3428 (+0.2080%)  0 (0.0000%)  -4 (-0.0015%)  3428 (+0.2077%)                                  312296
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
definitely needs to be disabled in minimize include and build_options/extract_featues files added

**Comment by IamPete1** [comment]
If scripting has to set it in anycase why not use scripting for the whole thing. You can receive and send MAVLink messages from Lua.

**Comment by nexton-winjeel** [comment]
@IamPete1: 
> If scripting has to set it in anycase why not use scripting for the whole thing? You can receive and send MAVLink messages from Lua.

Two reasons:
1. At least for the `CAMERA_INFORMATION` message, the `AP_Camera` driver is already sending this message. I didn't want a situation where the `AP_Camera` driver is sending the message and the scripting is sending a conflicting message.
2. These messages are requested rather than streamed. Is there a nice way to handle this in a script, rather than just periodically sending the message?

**Review by Hwurzburg — APPROVED** [review]
needs squash and library split

**Comment by nexton-winjeel** [comment]
@Hwurzburg:
> needs squash and library split

What do you mean?
* the commits are already split per library.
* the commits are logical changes that add one thing at a time, rather than a number of unrelated changes in a single commit.

**Comment by robertlong13** [comment]
> 2. These messages are requested rather than streamed. Is there a nice way to handle this in a script, rather than just periodically sending the message?

There is a way to do this, I actually did exactly that and was about to open a PR for it, but I also was thinking about redoing it to something like your approach here. It's easy enough to intercept a `MAV_CMD_REQUEST_VIDEO_STREAM_INFORMATION`, but that's being deprecated in favor of `MAV_CMD_REQUEST_MESSAGE`. The issue with the latter is that only one message request is cached for the lua script to grab, so I had to run the script at high speed to make sure I didn't miss a message request if the GCS sent a bunch of them all at once.

So, I think there are some advantages to your approach.

My lua script for reference
[video_stream.zip](https://github.com/user-attachments/files/16556947/video_stream.zip)

**Comment by Hwurzburg** [comment]
only one commit per library is allowed in final pr

**Comment by tpwrules** [comment]
> only one commit per library is allowed in final pr

This is definitely not true and IMO reduces clarity.

**Review by peterbarker — CHANGES_REQUESTED** [review]
> only one commit per library is allowed in final pr

That's not the case.  But there are too many commits in here.

This is a problem, 'though, @nexton-winjeel 
image

I usually also squash the changes to build_options.py and extract_features.py into a single commit (and often just omit the fact it touches extract_features.py) - there are two examples of that in the current commit list.

I'd also squash these two:
image

These could *possibly* also be squashed:
image

... they can be separated, but it's not particularly useful (and this, I think, the sort of thing that Henry might be kind of objecting to).

**Comment by nexton-winjeel** [comment]
@Hwurzburg, @peterbarker: Squashed.

**Comment by nexton-winjeel** [comment]
@peterbarker: Is `FenceFloorEnabledLanding` a flapping test? No code change, but it's now failing?

**Comment by andyp1per** [comment]
Its a new test - I haven't seen it flap

**Comment by nexton-winjeel** [comment]
I can't run tests, so I've rebased on `master` again to trigger a rebuild.

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 1]
```diff
@@ -769,6 +769,36 @@ singleton AP_Mount method get_angle_target boolean uint8_t'skip_check float'Null
...
+userdata mavlink_camera_information_t field sensor_size_v float'skip_check read write
+userdata mavlink_camera_information_t field flags uint32_t'skip_check read write
+userdata mavlink_camera_information_t field resolution_h uint16_t'skip_check read write
+userdata mavlink_camera_information_t field resolution_v uint16_t'skip_check read write
+userdata mavlink_camera_information_t field cam_definition_version uint16_t'skip_check read write
+userdata mavlink_camera_information_t field vendor_name'array 32 uint8_t'skip_check read write
+userdata mavlink_camera_information_t field model_name'array 32 uint8_t'skip_check read write
+userdata mavlink_camera_information_t field lens_id uint8_t'skip_check read write
+userdata mavlink_camera_information_t field cam_definition_uri'array 140 uint8_t'skip_check read write
+userdata mavlink_camera_information_t field gimbal_device_id uint8_t'skip_check read write
```
Comment (tridge) [inline libraries/AP_Scripting/generator/description/bindings.desc:786]: we should bring in the XML for camera_id

Comment (nexton-winjeel) [reply libraries/AP_Scripting/generator/description/bindings.desc:789]: @tridge: Is this PR blocked on getting the MAVLink XML changes for `camera_id` brought in?

**Comment by nexton-winjeel** [comment]
Looks like this is now overflowing the heap on the `CubeOrange-EKF2` build.

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 2]
```diff
@@ -772,6 +772,36 @@ singleton AP_Mount method get_angle_target boolean uint8_t'skip_check float'Null
...
+
+userdata mavlink_video_stream_information_t depends AP_CAMERA_ENABLED
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:791]:
```suggestion
userdata mavlink_video_stream_information_t depends AP_CAMERA_ENABLED && (AP_CAMERA_SCRIPTING_ENABLED == 1)
```

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 3]
```diff
@@ -772,6 +772,36 @@ singleton AP_Mount method get_angle_target boolean uint8_t'skip_check float'Null
 singleton AP_Mount method get_location_target boolean uint8_t'skip_check Location'Null
 singleton AP_Mount method set_attitude_euler void uint8_t'skip_check float'skip_check float'skip_check float'skip_check
 
+userdata mavlink_camera_information_t depends AP_CAMERA_ENABLED
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:775]:
```suggestion
userdata mavlink_camera_information_t depends AP_CAMERA_ENABLED && (AP_CAMERA_SCRIPTING_ENABLED == 1)
```

File: libraries/AP_Scripting/examples/set_VIDEO_STREAM_INFORMATION.lua [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    end
+    for i = 0, #uri do
+        stream_info:uri(i, uri:byte(i+1))
+    end
+
+    camera:set_stream_information(INSTANCE, stream_info)
+
+end
+
+return set_video_stream_information()
```
Comment (IamPete1) [inline libraries/AP_Scripting/examples/set_VIDEO_STREAM_INFORMATION.lua:33]:
```suggestion
return set_video_stream_information()

```

**Review by IamPete1 — CHANGES_REQUESTED** [review]
Depends should match the `AP_Camera` one.

**Comment by IamPete1** [comment]
Also the `AP_CAMERA_INFO_FROM_SCRIPT_ENABLED` `AP_MAVLINK_MSG_VIDEO_STREAM_INFORMATION_ENABLED` might need to be in the scripting bindings too? You can do a depends on methods too if you want the main scripting support.

**Comment by nexton-winjeel** [comment]
@IamPete1: I've updated the Lua bindings for the new structs and methods to depend on `AP_CAMERA_INFO_FROM_SCRIPT_ENABLED`.

**Comment by nexton-winjeel** [comment]
Rebased on `master` to get the changes from #28128 so CI passes.

**Comment by amilcarlucas** [comment]
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   3512 (+0.2622%)  0 (0.0000%)  0 (0.0000%)    3512 (+0.2618%)                                  620996
arducopter       3428 (+0.1896%)  0 (0.0000%)  4 (+0.0015%)   3428 (+0.1892%)                                  150640
ardusub          3428 (+0.2142%)  0 (0.0000%)  4 (+0.0015%)   3428 (+0.2139%)                                  359756
blimp            3512 (+0.2579%)  0 (0.0000%)  0 (0.0000%)    3512 (+0.2575%)                                  598928
arduplane        3428 (+0.1899%)  0 (0.0000%)  4 (+0.0015%)   3428 (+0.1895%)                                  153320
ardurover        3432 (+0.2070%)  0 (0.0000%)  0 (0.0000%)    3432 (+0.2067%)                                  302104
arducopter-heli  3428 (+0.1897%)  0 (0.0000%)  -4 (-0.0015%)  3428 (+0.1893%)                                  151600
```

It's still big, but if the ifdefs work now, it should not be a problem.

**Comment by nexton-winjeel** [comment]
Running locally, I get the following:

This branch, `AP_CAMERA_INFO_FROM_SCRIPT_ENABLED = 0`:
```
Build directory: /ardupilot/build/sitl
Target          Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
----------------------------------------------------------------------------------------------------------
bin/arducopter   4172858    191517   222752               4364375  Not Applicable  Not Applicable
```

`master` branch:
```
Build directory: /ardupilot/build/sitl
Target          Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
----------------------------------------------------------------------------------------------------------
bin/arducopter   4172978    191453   222752               4364431  Not Applicable  Not Applicable  
```

Deltas between disabled and `master`:
```
Target          Text (B)  Data (B)  BSS (B)  Total Flash Used (B)  Free Flash (B)  External Flash Used (B)
----------------------------------------------------------------------------------------------------------
bin/arducopter      -120        64        0                   -56  Not Applicable  Not Applicable  
```

**Comment by rmackay9** [comment]
@Hwurzburg, I think the requirement is that individual commits should not affect more than one subsystem but multiple commits affecting a single subsystem is OK.  The git-subsystems-split script is probably meant to divide up a single commit but not to consolidate commits.

File: Tools/scripts/build_options.py [hunk 5]
```diff
@@ -141,6 +141,7 @@ def config_option(self):
     Feature('Camera', 'Camera_Solo', 'AP_CAMERA_SOLOGIMBAL_ENABLED', 'Enable Solo Gimbal support', 0, 'Camera'),
     Feature('Camera', 'Camera_FOV_Status', 'AP_CAMERA_SEND_FOV_STATUS_ENABLED', 'Enable Camera FOV Status send to GCS', 0, 'Camera,MOUNT'),  # noqa: E501
     Feature('Camera', 'Camera_ThermalRange', 'AP_CAMERA_SEND_THERMAL_RANGE_ENABLED', 'Enable Camera Thermal Range send to GCS', 0, 'Camera,MOUNT'),  # noqa: E501
+    Feature('Camera', 'Camera_Info_From_Script', 'AP_CAMERA_INFO_FROM_SCRIPT_ENABLED', 'Enable setting Camera information messages from Lua script', 0, 'Camera,SCRIPTING'), # noqa
```
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply Tools/scripts/build_options.py:144]: should actually be( to conform to style):
'Enable Camera information messages via Lua script'

File: libraries/AP_Camera/AP_Camera_Backend.cpp [hunk 6]
```diff
@@ -212,31 +223,61 @@ void AP_Camera_Backend::send_camera_feedback(mavlink_channel_t chan)
...
-        get_gimbal_device_id());// gimbal_device_id uint8_t
+    camera_info = _camera_info;
+
+#else
+
+    memset(&camera_info, 0, sizeof(camera_info));
+
+    camera_info.focal_length = NAN;
+    camera_info.sensor_size_h = NAN;
+    camera_info.sensor_size_v = NAN;
```
[the remark this answers is not in the document]
Comment (nexton-winjeel) [reply libraries/AP_Camera/AP_Camera_Backend.cpp]: Fixed.

File: Tools/scripts/build_options.py [hunk 7]
```diff
@@ -370,6 +371,7 @@ def config_option(self):
     Feature('MAVLink', 'MAVLINK_MSG_RC_CHANNELS_RAW', 'AP_MAVLINK_MSG_RC_CHANNELS_RAW_ENABLED', 'Enable sending of RC_CHANNELS_RAW mavlink messages', 0, None),  # noqa
     Feature('MAVLink', 'AP_MAVLINK_FTP_ENABLED', 'AP_MAVLINK_FTP_ENABLED', 'Enable MAVLink FTP Protocol', 0, None),  # noqa
     Feature('MAVLink', 'MAV_CMD_SET_HAGL', 'AP_MAVLINK_MAV_CMD_SET_HAGL_ENABLED', 'Enable MAVLink HAGL command', 0, None),  # noqa
+    Feature('MAVLink', 'VIDEO_STREAM_INFORMATION', 'AP_MAVLINK_MSG_VIDEO_STREAM_INFORMATION_ENABLED', 'Enable MAVLink AP_MAVLINK_MSG_VIDEO_STREAM_INFORMATION_ENABLED message', 0, "Camera"), # noqa
```
[the remark this answers is not in the document]
Comment (nexton-winjeel) [reply Tools/scripts/build_options.py]: Fixed.

File: Tools/scripts/build_options.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nexton-winjeel) [reply Tools/scripts/build_options.py]: Fixed with the wording from @Hwurzburg.

**Review by rmackay9 — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]

**Pull request merged**